FreeRTOS-Plus/Demo/AWS/Device_Defender_Windows_Simulator/Device_Defender_Demo/metrics_collector.c

```diff
@@ -0,0 +1,243 @@
+        pxOutNetworkStats->ulPacketsReceived = xMetrics.xInput.uxPacketCount;
+        pxOutNetworkStats->ulBytesSent = xMetrics.XOutput.uxByteCount;
+        pxOutNetworkStats->ulPacketsSent = xMetrics.XOutput.uxPacketCount;
+    }
+
+    return eStatus;
+}
+/*-----------------------------------------------------------*/
+
+MetricsCollectorStatus_t xGetOpenTcpPorts( uint16_t * pusOutTcpPortsArray,
+                                           uint32_t ulTcpPortsArrayLength,
+                                           uint32_t * pulOutNumTcpOpenPorts )
+{
+    MetricsCollectorStatus_t status = MetricsCollectorSuccess;
+
+    MetricsType_t xMetrics = { 0 };
+    BaseType_t xMetricsStatus = 0;
+
+    if( pulOutNumTcpOpenPorts == NULL )
+    {
+        LogError( ( "Invalid parameter. pulOutNumTcpOpenPorts NULL." ) );
+        status = MetricsCollectorBadParameter;
+    }
+
+    if( status == MetricsCollectorSuccess )
+    {
+        xMetricsStatus = vGetMetrics( &xMetrics );
+
+        if( xMetricsStatus != 0 )
+        {
+            status = MetricsCollectorCollectionFailed;
+        }
+    }
+
+    if( status == MetricsCollectorSuccess )
+    {
+        *pulOutNumTcpOpenPorts = xMetrics.xTCPPortList.uxCount;
+
+        if( pusOutTcpPortsArray != NULL )
+        {
+            if( xMetrics.xTCPPortList.uxCount < ulTcpPortsArrayLength )
+            {
+                ulTcpPortsArrayLength = xMetrics.xTCPPortList.uxCount;
+            }
+
+            memcpy( pusOutTcpPortsArray, &xMetrics.xTCPPortList.usTCPPortList, ulTcpPortsArrayLength * sizeof( uint16_t ) );
+        }
+    }
+
+    return MetricsCollectorSuccess;
+}
+/*-----------------------------------------------------------*/
+
+MetricsCollectorStatus_t xGetOpenUdpPorts( uint16_t * pusOutUdpPortsArray,
+                                           uint32_t ulUdpPortsArrayLength,
+                                           uint32_t * pulOutNumUdpOpenPorts )
+{
+    MetricsCollectorStatus_t eStatus = MetricsCollectorSuccess;
+
+    MetricsType_t xMetrics = { 0 };
+    BaseType_t xMetricsStatus = 0;
+
+    if( pulOutNumUdpOpenPorts == NULL )
+    {
+        LogError( ( "Invalid parameter. pOutNumudpOpenPorts NULL." ) );
+        eStatus = MetricsCollectorBadParameter;
+    }
+
+    if( eStatus == MetricsCollectorSuccess )
+    {
+        xMetricsStatus = vGetMetrics( &xMetrics );
+
+        if( xMetricsStatus != 0 )
+        {
+            eStatus = MetricsCollectorCollectionFailed;
+        }
+    }
+
+    if( eStatus == MetricsCollectorSuccess )
+    {
+        *pulOutNumUdpOpenPorts = xMetrics.xTCPPortList.uxCount;
+
+        if( pusOutUdpPortsArray != NULL )
+        {
+            if( xMetrics.xUDPPortList.uxCount < ulUdpPortsArrayLength )
+            {
+                ulUdpPortsArrayLength = xMetrics.xUDPPortList.uxCount;
+            }
+
+            memcpy( pusOutUdpPortsArray, &xMetrics.xUDPPortList.usUDPPortList, ulUdpPortsArrayLength * sizeof( uint16_t ) );
```

Comment:
Same question as in the case of TCP about if xMetrics.xUDPPortList.uxCount > ulUdpPortsArrayLength